### PR TITLE
1815 mp content files versioning bug

### PR DESCRIPTION
**Pull request by jsadler2**
@pkdash could you please review this?

### CODE REVIEW

File: hs_model_program/forms.py [hunk 1]
```diff
@@ -12,13 +12,12 @@ class mp_form_helper(BaseFormHelper):
     def __init__(self, allow_edit=True, res_short_id=None, element_id=None, element_name=None, *args, **kwargs):
 
         files = kwargs.pop('files')
-        file_data = {}
+        file_names = []
         for f in files.all():
             short_url = escape(f.resource_file.name)
-            name = short_url.split('/')[-1]
-            file_data[name] = short_url
+            file_names.append(short_url.split('/')[-1])
```
Comment (pkdash) [inline hs_model_program/forms.py]: You can generate the list of file names like this:

```file_names.append(f.file_name)```

File: hs_model_program/templates/pages/modelprogramresource.html [hunk 2]
```diff
@@ -13,59 +13,58 @@
...
                     {########################}
-
-                    {% if mpmetadata.get_engine_list.items|first|first or mpmetadata.get_software_list.items|first|first or mpmetadata.get_releasenotes_list.items|first|first or mpmetadata.get_documentation_list.items|first|first %}
+                    {% if mpmetadata.get_engine_list or mpmetadata.get_software_list or mpmetadata.get_releasenotes_list or mpmetadata.get_documentation_list %}
                     <legend>Content files</legend>
                     <table class="custom-table">
                         <tbody>
-                        {% if mpmetadata.get_engine_list.items|first|first %}
-                            {% for name, path in mpmetadata.get_engine_list.items %}
+                        {% if mpmetadata.get_engine_list %}
```
Comment (pkdash) [inline hs_model_program/templates/pages/modelprogramresource.html]: This ```if``` is not needed.

File: hs_model_program/templates/pages/modelprogramresource.html [hunk 3]
```diff
@@ -13,59 +13,58 @@
...
-                                           href="http://{{ request.get_host }}/resource/{{ path }}"
+                                           href="http://{{ request.get_host }}/hsapi/resource/{{ cm.short_id }}/files/{{ name }}"
                                            target="_blank">{{ name }}</a></td>
                                 </tr>
                             {% endfor %}
                         {% endif %}
 
-                        {% if mpmetadata.get_software_list.items|first|first %}
-                            {% for name, path in mpmetadata.get_software_list.items %}
+                        {% if mpmetadata.get_software_list %}
```
Comment (pkdash) [inline hs_model_program/templates/pages/modelprogramresource.html]: Same comment as above

File: hs_model_program/templates/pages/modelprogramresource.html [hunk 4]
```diff
@@ -13,59 +13,58 @@
...
-                                           href="http://{{ request.get_host }}/resource/{{ path }}"
+                                           href="http://{{ request.get_host }}/hsapi/resource/{{ cm.short_id }}/files/{{ name }}"
                                            target="_blank">{{ name }}</a></td>
                                 </tr>
                             {% endfor %}
                         {% endif %}
 
-                        {% if mpmetadata.get_documentation_list.items|first|first %}
-                            {% for name, path in mpmetadata.get_documentation_list.items %}
+                        {% if mpmetadata.get_documentation_list %}
```
Comment (pkdash) [inline hs_model_program/templates/pages/modelprogramresource.html]: Same comment as above

File: hs_model_program/templates/pages/modelprogramresource.html [hunk 5]
```diff
@@ -13,59 +13,58 @@
...
+                                           href="http://{{ request.get_host }}/hsapi/resource/{{ cm.short_id }}/files/{{ name }}"
                                            target="_blank">{{ name }}</a></td>
                                 </tr>
                             {% endfor %}
                         {% endif %}
 
 
-                        {% if mpmetadata.get_releasenotes_list.items|first|first %}
-                            {% for name, path in mpmetadata.get_releasenotes_list.items %}
+                        {% if mpmetadata.get_releasenotes_list %}
```
Comment (pkdash) [inline hs_model_program/templates/pages/modelprogramresource.html]: Same as above.

File: hs_model_program/views.py [hunk 6]
```diff
@@ -37,10 +37,10 @@ def get_model_metadata(request):
                 software_language=mpmeta.modelProgramLanguage,
                 operating_sys=mpmeta.modelOperatingSystem,
                 url = protocol+"://"+request.get_host()+"/resource/"+resource_id+"/",
-                modelEngine = mpmeta.modelEngine.split(';'),
-                modelSoftware=mpmeta.modelSoftware.split(';'),
-                modelDocumentation=mpmeta.modelDocumentation.split(';'),
-                modelReleaseNotes=mpmeta.modelReleaseNotes.split(';'),
+                modelEngine = mpmeta.get_engine_list(),
```
Comment (pkdash) [inline hs_model_program/views.py]: Where are these new context variables used in the template?

**Comment by pkdash** [comment]
@jsadler2 I have left few review comments.

File: hs_model_program/forms.py (hunk 1 above)
Comment (jsadler2) [inline hs_model_program/forms.py]: Thanks. That's changed now. Much simpler

File: hs_model_program/templates/pages/modelprogramresource.html (hunk 2 above)
Comment (jsadler2) [inline hs_model_program/templates/pages/modelprogramresource.html]: On these ones, if I don't include the `if`, the row with the heading shows up even when there are no files: 
image

Actually, when I was testing it, what I thought was working `{% if mpmetadata.get_engine_list %}` did not work and it needed to be `{% if mpmetadata.get_engine_list|first %}` resulting in what we want: 
image

File: hs_model_program/templates/pages/modelprogramresource.html (hunk 3 above)
Comment (jsadler2) [inline hs_model_program/templates/pages/modelprogramresource.html]: see above

File: hs_model_program/templates/pages/modelprogramresource.html (hunk 4 above)
Comment (jsadler2) [inline hs_model_program/templates/pages/modelprogramresource.html]: see above

File: hs_model_program/templates/pages/modelprogramresource.html (hunk 5 above)
Comment (jsadler2) [inline hs_model_program/templates/pages/modelprogramresource.html]: see above

File: hs_model_program/views.py (hunk 6 above)
Comment (jsadler2) [inline hs_model_program/views.py]: This view is called via ajax in the javascript to populate the multi-select elements in the resource edit mode: https://github.com/hydroshare/hydroshare/blob/c75c8dda3fa3fe73fff85e8c4c7a91abc3afb0e5/hs_model_program/static/js/model_program.js#L51

**Comment by jsadler2** [comment]
@pkdash I have responded to your review comments

File: hs_model_program/views.py (hunk 6 above)
Comment (pkdash) [inline hs_model_program/views.py]: @jsadler2 I am not seeing for example where in the javascript the context variable ```modelEngine``` is being used.

File: hs_model_program/templates/pages/modelprogramresource.html [hunk 7]
```diff
@@ -13,59 +13,58 @@
...
                     {########################}
-
-                    {% if mpmetadata.get_engine_list.items|first|first or mpmetadata.get_software_list.items|first|first or mpmetadata.get_releasenotes_list.items|first|first or mpmetadata.get_documentation_list.items|first|first %}
+                    {% if mpmetadata.get_engine_list|first or mpmetadata.get_software_list|first or mpmetadata.get_releasenotes_list|first  or mpmetadata.get_documentation_list|first  %}
                     <legend>Content files</legend>
                     <table class="custom-table">
                         <tbody>
-                        {% if mpmetadata.get_engine_list.items|first|first %}
-                            {% for name, path in mpmetadata.get_engine_list.items %}
+                        {% if mpmetadata.get_engine_list|first %}
```
Comment (pkdash) [inline hs_model_program/templates/pages/modelprogramresource.html]: Why do we need this if condition?

File: hs_model_program/views.py (hunk 6 above)
Comment (jsadler2) [inline hs_model_program/views.py]: @pkdash in the javascript, each of the multiselect elements is looped through (starting line 58). These multiselect boxes are for the different model file metadata elements (i.e. `modelEngine`, `modelDocumentation`, `modelReleaseNotes`, `modelEngine`). The corresponding files names stored under these terms come from the AJAX  data (line 64). Then the file names are added to the multiselect (starting in line 68).

File: hs_model_program/templates/pages/modelprogramresource.html (hunk 7 above)
Comment (jsadler2) [inline hs_model_program/templates/pages/modelprogramresource.html]: Without this `if` condition, the "Model Engine" (or "Model Software" etc.) heading will appear without any file names there. For example, 
image

Comment (pkdash) [inline hs_model_program/templates/pages/modelprogramresource.html]: Got it.

File: hs_model_program/views.py (hunk 6 above)
Comment (pkdash) [inline hs_model_program/views.py]: My mistake. I thought I am looking at the page_processor.py file. That's why I was referring to these variables as context variables.

**Comment by pkdash** [comment]
+1

**Pull request merged**